### PR TITLE
#161966622  Search functionality and custom filtering

### PR DESCRIPTION
**Pull request by salma-nyagaka**
### What does this PR do?
This PR adds the functionality of searching and filtering of articles

###### Use the following endpoints to test the functionalities that have been implemented.

| Method | Endpoint                                          | Description                                                                        | 
| ------    | -------------------------------    | -------------------------------------                             | 
| GET       | /api/article/?author={'username'} | Filter by author's username                                              |
| GET       | /api/article/?title={'title'}                | Filter by article's title                                                        | 
| GET       | /api/article/?tag={'tag'}                 | Filter by tags                                                                      |
| GET       | /api/article/?search={'text'}          | Seartch by title, author and tag                                        |                                                     
### How should this be manually tested

- Clone the GitHub repository `git clone https://github.com/andela/ah-technocrats`
- Checkout to the branch `git checkout ft-search-filter-161966622`
- Create a virtual environment `virtualenv venv`
- Install the dependencies `pip install -r requiremnts.txt`
- Run migrations `python manage.py makemigrations`  `python manage.py migrate`
- Test the application `python manage.py test`
- Test with postman after running `python manage.py runserver`

### Any background context you want to provide?

A user can filter articles by the tag, title and author 
A user can be able to search for an article by the tag, title and author

What are the relevant pivotal tracker stories?
#161966622(https://www.pivotaltracker.com/story/show/161966622)
# Checklist:
- [x] My code follows the style guidelines of this project
- [ ] At least 2 people have reviewed my PR
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] My PR has one commit.

#### Filter by title
filter title

#### Filter by author's username
filter author

###filter by tag
screen shot 2018-12-20 at 11 29 14

#### Response if search is not found

screen shot 2018-12-20 at 22 10 30

#### Search by the title, author and tag
search title

### CODE REVIEW

**Comment by Mnickii** [comment]
Let's have GET	/api/article/view/?tag={'tag'}	Filter by tags be 	/api/article/?tag=AngularJs

Look at the README for the desired endpoint given.

Query Parameters:

Filter by tag:

?tag=AngularJS

**Review by Mnickii — DISMISSED** [review]
Left a comment. Look at the README for the articles endpoint for the desired filter endpoints.

File: authors/apps/articles/tests/test_search.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+from django.urls import reverse
+from rest_framework import status
+from .base_test import BaseTestCase
+
+
+class TestSearchArticles(BaseTestCase):
+    "class to test searching and filtering of articles"
```
Comment (SnyderMbishai) [inline authors/apps/articles/tests/test_search.py]: Check on the docstring. Should be three pairs of quotes like the other ones.

**Review by SnyderMbishai — DISMISSED** [review]

**Comment by evansmusomi** [comment]
> Let's have GET	/api/article/view/?tag={'tag'}	Filter by tags be /api/article/?tag=AngularJs
> 
> Look at the README for the desired endpoint given.
> 
> Query Parameters:
> 
> Filter by tag:
> 
> ?tag=AngularJS

@Mnickii, I've confirmed this change has been made.

**Review by verenceLola — APPROVED** [review]
Nice work @salma-nyagaka

**Review by TeamoreA — APPROVED** [review]

**Review by evansmusomi — APPROVED** [review]
The updated version looks good to me. Changes requested have been incorporated.

**Pull request merged**